framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,601 @@
+            String basename,
+            @FullyQualifiedName String classname,
+            AClass aClass,
+            PrintWriter printWriter) {
+
+        String nameToPrint = basename;
+        String remainingInnerClassNames = "";
+        if (basename.contains(".")) {
+            nameToPrint = basename.substring(0, basename.indexOf('.'));
+            remainingInnerClassNames = basename.substring(basename.indexOf('.') + 1);
+        }
+
+        // For any outer class, print "class".  For a leaf class, print "enum" or "class".
+        if ("".equals(remainingInnerClassNames) && enumConstants.containsKey(classname)) {
+            printWriter.print("enum ");
+        } else {
+            printWriter.print("class ");
+        }
+        formatAnnotations(aClass.tlAnnotationsHere);
+        printWriter.print(nameToPrint);
+        printTypeParameters(classname, printWriter);
+        printWriter.println(" {");
+        if ("".equals(remainingInnerClassNames)) {
+            return 0;
+        } else {
+            return 1
+                    + printClassDefinitions(
+                            remainingInnerClassNames, classname, aClass, printWriter);
+        }
+    }
+
+    /**
+     * Prints all the fields of a given class
+     *
+     * @param aClass the class whose fields should be printed
+     * @param fullyQualifiedClassname the fully-qualified name of the class
+     * @param printWriter the writer on which to print the fields
+     */
+    private void printFields(
+            AClass aClass,
+            @FullyQualifiedName String fullyQualifiedClassname,
+            PrintWriter printWriter) {
+        for (Map.Entry<String, AField> fieldEntry : aClass.fields.entrySet()) {
+            String fieldName = fieldEntry.getKey();
+            AField aField = fieldEntry.getValue();
+            printWriter.println();
+            printWriter.print(INDENT);
+            printWriter.print(formatAField(aField, fieldName, fullyQualifiedClassname));
+            printWriter.println(";");
+        }
+    }
+
+    /**
+     * Prints a method signature in stub file format (i.e., without a method body).
+     *
+     * @param aMethod the method to print
+     * @param basename the simple name of the containing class. Used only to determine if the method
+     *     being printed is the constructor of an inner class.
+     * @param fullyQualifiedClassname the fully-qualified name of the containing class
+     * @param printWriter where to print the method signature
+     */
+    private void printMethodSignature(
+            AMethod aMethod,
+            String basename,
+            @FullyQualifiedName String fullyQualifiedClassname,
+            PrintWriter printWriter) {
+        printWriter.println();
+        printWriter.print(INDENT);
+        printWriter.print(formatAnnotations(aMethod.returnType.tlAnnotationsHere));
+        String methodName = aMethod.methodName.substring(0, aMethod.methodName.indexOf("("));
```

Comment:
Why does a method name contain "("?  I suspect this call to `substring` can be removed.  My perusal of `AMethod` seems to confirm that.

Reply:
This isn't correct - AMethod's `methodName` field contains the name of the method and the types of each parameter, to distinguish between different methods in the same class that share a name. I added a comment.

Reply:
For example, a method `foo` that takes a String and returns a collection is represented in the `methodName` field as `foo(Ljava/lang/String;)Ljava/util/Collection;`